File: src_new/gen_dat.py
import struct
import sys

OF_NUM = 5
STRIP_NUM = 2
VERSION_MAJOR = 1
VERSION_MINOR = 0
FRAME_NUM = 5
LED_num_array = [10, 20]
# 範例，這是版本1.0，有五條光纖，兩條燈條上面有{10, 20}顆LED，總共5個frame

def calculate_checksum(frame_data):
    #計算checksum=所有byte的和 mod 2^32
    return sum(frame_data) & 0xFFFFFFFF

def main():
    total_leds = sum(LED_num_array)
    
    frame_size_without_checksum = 4+1+(OF_NUM *3)+(total_leds *3)
    frame_size_with_checksum = frame_size_without_checksum+4
    
    #1.生control.dat
    with open("control.dat", "wb") as control_file:
        control_file.write(struct.pack('<BB', VERSION_MAJOR, VERSION_MINOR))
        control_file.write(struct.pack('<B', OF_NUM))
        control_file.write(struct.pack('<B', STRIP_NUM))
        
        for led_num in LED_num_array:
            control_file.write(struct.pack('<B', led_num))

        control_file.write(struct.pack('<I', FRAME_NUM))
        
        for k in range(FRAME_NUM):
            timestamp = k * 100
            control_file.write(struct.pack('<I', timestamp))
            # 範例，第k frame的timestamp是100*k
    
    print("control.dat finish")
    
    #2.生frame.dat
    with open("frame.dat", "wb") as frame_file:
        frame_file.write(struct.pack('<BB', VERSION_MAJOR, VERSION_MINOR))
    
        for k in range(FRAME_NUM):
            frame_data = bytearray()
            start_time = k * 100
            frame_data.extend(struct.pack('<I', start_time))
            fade = 1
            frame_data.append(fade)
            
            # 範例，第k frame的OF GRB生成邏輯是: OF[i].G/R/B = (i+1/2/3 + k) mod 255
            for i in range(OF_NUM):
                # G, R, B 各 1 byte
                frame_data.append((i + 1 + k) % 255)  # G
                frame_data.append((i + 2 + k) % 255)  # R
                frame_data.append((i + 3 + k) % 255)  # B
            
            # 範例，第k frame的LED GRB生成邏輯是: LED[i][j].G/R/B = ((i*10+j) + 1/2/3 + k) mod 255
            for i in range(STRIP_NUM):
                for j in range(LED_num_array[i]):
                    base = i * 10 + j
                    frame_data.append((base + 1 + k) % 255)  # G
                    frame_data.append((base + 2 + k) % 255)  # R
                    frame_data.append((base + 3 + k) % 255)  # B
            
            checksum = calculate_checksum(frame_data)
            frame_file.write(frame_data)
            frame_file.write(struct.pack('<I', checksum))
    
    print("\nframe.dat finish")
    print(f"OF_num: {OF_NUM}")
    print(f"Strip_num: {STRIP_NUM}")
    print(f"LED_num: {' '.join(map(str, LED_num_array))}")
    print(f"total LED: {total_leds}")
    print(f"Version: {VERSION_MAJOR}.{VERSION_MINOR}")
    print(f"Frame_num: {FRAME_NUM}")
    print(f"frame_size with checksum: {frame_size_with_checksum} byte")
    print()

if __name__ == "__main__":
    main()

File: src_new/gen_dat.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define OF_NUM 10
#define STRIP_NUM 5
#define VERSION_MAJOR 1
#define VERSION_MINOR 3
#define FRAME_NUM 30
const uint8_t LED_num_array[STRIP_NUM] = {5, 10, 15, 20, 25};
//範例版本1.3，有10條光纖，5條燈條上面有{5, 10, 15, 20, 25}顆LED，總共30個frame

uint32_t calculate_checksum(uint8_t *frame_data, uint32_t frame_size_without_checksum){
    uint32_t sum = 0;
    for (uint32_t i = 0; i < frame_size_without_checksum; i++) {
        sum += frame_data[i];
    }
    return sum;
}

int main(){
    //算total led
    uint32_t total_leds = 0;
    for (int i = 0; i < STRIP_NUM; i++){
        total_leds += LED_num_array[i];
    }
    //算frame_size
    uint32_t frame_size_without_checksum = 4 + 1 + (OF_NUM * 3) + (total_leds * 3);
    uint32_t frame_size_with_checksum = frame_size_without_checksum + 4;
    
    //1.生control.dat
    FILE *control_file = fopen("control.dat", "wb");
    uint8_t version[2] = {VERSION_MAJOR, VERSION_MINOR};
    fwrite(version, 1, 2, control_file);
    
    uint8_t of_num = OF_NUM;
    uint8_t strip_num = STRIP_NUM;
    fwrite(&of_num, 1, 1, control_file);
    fwrite(&strip_num, 1, 1, control_file);
    
    for (int i = 0; i < STRIP_NUM; i++){
        fwrite(&LED_num_array[i], 1, 1, control_file);
    }
    
    uint32_t frame_num_le = FRAME_NUM;
    fwrite(&frame_num_le, 4, 1, control_file);
    
    for (int k = 0; k < FRAME_NUM; k++){
        uint32_t timestamp = k * 100;
        fwrite(&timestamp, 4, 1, control_file);
    }
    //範例，第k frame的timestamp是100*k
    fclose(control_file);
    printf("control.dat finish\n");
    
    //2.生frame.dat
    FILE *frame_file = fopen("frame.dat", "wb");
    
    fwrite(version, 1, 2, frame_file);
    
    //生成每個frame
    for (int k = 0; k < FRAME_NUM; k++){
        uint8_t *frame_data = malloc(frame_size_without_checksum);
        uint32_t offset = 0;
        uint32_t start_time = k * 100;
        memcpy(frame_data + offset, &start_time, 4);
        offset += 4;
        
        uint8_t fade = 1;
        frame_data[offset] = fade;
        offset += 1;
        
        //範例第k frame的OF GRB生成邏輯是: OF[i].G/R/B = (i+1/2/3 + k) mod 255
        for (int i = 0; i < OF_NUM; i++){
            frame_data[offset] = (i + 1 + k) % 255;
            frame_data[offset + 1] = (i + 2 + k) % 255;
            frame_data[offset + 2] = (i + 3 + k) % 255;
            offset += 3;
        }
        
        //範例第k frame的LED GRB生成邏輯是: LED[i][j].G/R/B = ((i*10+j) + 1/2/3 + k) mod 255
        for (int i = 0; i < STRIP_NUM; i++){
            for (int j = 0; j < LED_num_array[i]; j++){
                int base = i * 10 + j;
                frame_data[offset] = (base + 1 + k) % 255; // G
                frame_data[offset + 1] = (base + 2 + k) % 255; // R
                frame_data[offset + 2] = (base + 3 + k) % 255; // B
                offset += 3;
            }
        }
        
        uint32_t checksum = calculate_checksum(frame_data, frame_size_without_checksum);
        fwrite(frame_data, 1, frame_size_without_checksum, frame_file);
        fwrite(&checksum, 4, 1, frame_file);
        free(frame_data);
    }
    
    fclose(frame_file);

    printf("\n");
    printf("frame.dat finish\n");
    printf("OF_num: %d\n", OF_NUM);
    printf("Strip_num: %d\n", STRIP_NUM);
    printf("LED_num: ");
    for (int i = 0; i < STRIP_NUM; i++){
        printf("%d ", LED_num_array[i]);
    }
    printf("\n");
    printf("total LED: %u\n", total_leds);
    printf("Version: %d.%d\n", VERSION_MAJOR, VERSION_MINOR);
    printf("Frame_num: %d\n", FRAME_NUM);
    printf("frame_size with checksum: %u byte\n", frame_size_with_checksum);
    printf("\n");
    return 0;
}